### ISSUE TITLE
What makes fastForEach Faster

### ISSUE THREAD

**Issue by mshoote** [brianmhunt/knockout-fast-foreach#48]
Hi, I was wondering what exactly makes fast for each faster than the normal for each.

I noticed in the .js you mention:
"Employing sound techniques to make a faster Knockout foreach binding."

I am not sure if these sound techniques make it faster and if so what are sound techniques? 

I am wondering as I am doing a comparison of frameworks and was wondering what made this faster.

**Comment by brianmhunt** [comment]
Good question. :)  There are a few things, including (and I may be forgetting things):

1. Updates are batched into [animationFrames](https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame)

2. It has a `noIndex` option that makes it `O(1)` in terms of computational complexity

3. DOM nodes are re-used for re-ordering/deletion/inserts within the same animation frame

4. It can re-use the Knockout Context with a `noContext` option

So, in other words, in the average case it'll have fewer reflows because DOM changes are batched into animation frames.  When given options `noIndex` it will scale to relatively huge numbers because each addition has constant-time computational complexity.

I hope that helps explain. (It's also possible I've forgotten something, mind you ...)

**Comment by mshoote** [comment]
Thank you for the response! That helps a lot.

**Comment by krnlde** [comment]
So this issue can be closed?

**Comment by mshoote** [comment]
Yes this issue can be closed.
